apitools/base/py/transfer_test.py

```diff
@@ -477,3 +484,77 @@ def testMediaCompressed(self):
         with gzip.GzipFile(fileobj=self.request.body) as f:
             original = f.read()
             self.assertTrue(self.sample_data in original)
+
+    def HttpRequestSideEffect(self, responses=None):
+        responses = [(response.info, response.content) for response in responses]
+        def _side_effect(uri, **kwargs):  # pylint: disable=unused-argument
+            body = kwargs['body']
+            read_func = getattr(body, 'read', None)
+            if read_func:
+                # If the body is a stream, consume the stream.
+                body = read_func()
+            self.assertEqual(int(kwargs['headers']['content-length']), len(body))
+            return responses.pop()
+        return _side_effect
+
+    def testRetryRequestChunks(self):
+        """Test that StreamInChunks will retry correctly."""
+        # Create and configure the upload object.
+        bytes_http = httplib2.Http()
+        upload = transfer.Upload(
+            stream=self.sample_stream,
+            mime_type='text/plain',
+            total_size=len(self.sample_data),
+            close_stream=False,
+            http=bytes_http)
+
+        upload.strategy = transfer.RESUMABLE_UPLOAD
+        # Set the chunk size so the entire stream is uploaded.
+        upload.chunksize = len(self.sample_data)
+        # Mock the upload to return the sample response.
+        with mock.patch.object(bytes_http,
+                               'request') as make_request:
+            # This side effect also checks the request body.
+            make_request.side_effect = self.HttpRequestSideEffect([
+                self.response,  # Initial request in InitializeUpload().
+                self.fail_response,  # 503 status code from server.
+                self.response # Successful request.
+            ])
+
+            # Initialization.
+            upload.InitializeUpload(self.request, bytes_http)
+            upload.StreamInChunks()
+
+            # Ensure the mock was called three times
```

Comment:
Why three? A short comment may be helpful.

Also, would it be useful to assert anything else?

Reply:
Do you think another comment is necessary here. In the side_effect above, I've commented on the what the responses are.

Reply:
Ah ok. May be extract responses into a local variable and use the len of that in the assert?

Reply:
Done.